File: t14.py
# -*- conding:utf-8 -*-
#Author:lyc
import os, sys

# 题目：
# 编写一个函数来查找字符串数组中的最长公共前缀。
# 如果不存在公共前缀，返回空字符串 ""。

# 示例 1:
# 输入: ["flower","flow","flight"]
# 输出: "fl"

# 示例 2:
# 输入: ["dog","racecar","car"]
# 输出: ""
# 解释: 输入不存在公共前缀。

# 说明:
# 所有输入只包含小写字母 a-z 。



class Solution:
    def longestCommonPrefix(self, strs: str) -> str:
        """
        遍历每一个列表元素取出对应的下标元素去重处理
        :param strs:
        :return:
        """

        if len(strs) == 0:
            return ""

        min_s = len(min(strs))

        result = ""

        for i in range(min_s):
            i_list = []
            for j in strs:
                i_list.append(j[i])
            if len(set(i_list)) == 1:
                result += i_list[0]
            else:
                break
        return result


result = Solution()
#x = ["flower","flow","flight"]
x = []
print(result.longestCommonPrefix(x))

File: C++/t14.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <set>

using namespace std;

/**
 *  leetcode第14题
 *  寻找vector中最短元素，遍历，每次按照下标构建一个vector去重看长度是否等于1
 *  Author:lyc
*/


class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {

        int strs_size = strs.size();

        // 空vector直接返回
        if(strs.size() == 0){
            return "";
        };

        // 创建结果字符串
        string result = "";

        // 寻找最短元素下标及长度
        int minPosition = max_element(strs.begin(),strs.end()) - strs.begin();
        int minLength = strs[minPosition].length();

        for(int i=0; i < minLength; i++){

            // 构建临时vector
            vector<char> v {};
            for(int j = 0; j < strs.size(); j++){
                v.push_back(strs[j][i]);
            }

            // 去重
            set<char> s(v.begin(), v.end());
            v.assign(s.begin(), s.end());

            // 判断去重之后是否长度为1
            if(v.size() == 1){
                result += v[0];
            }else{
                break;
            }
        }

        return result;

    }
};

int main()
{
    Solution model;

    vector<string> x {"flower","flow","flight"};

    string res = model.longestCommonPrefix(x);
    cout << res << endl;

    system("pause");
    return 0;
}



